plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/DevWatchBootableJarMojo.java

```diff
@@ -293,6 +338,131 @@ ModelNode waitStatus(ModelControllerClient client, String status, ModelNode op)
         }
     }
 
+    private class RemoteDeploymentController implements DeploymentController {
+        private final String name;
+
+        public RemoteDeploymentController(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public void deploy(Path dir) throws Exception {
+            try (ModelControllerClient client = createClient()) {
+                ServerHelper.waitForStandalone(client, timeout);
+                undeploy(client, name);
+                waitRemoved(client, name);
+                boolean success = deploy(client, dir);
+                if (success) {
+                    waitDeploymentUp(client, name);
+                }
+                // We only need this on Windows since it may lock the JAR when the delete process is running
+                if (IS_WINDOWS) {
+                    currentServerDir = getHomeDirectory(client);
+                }
+            }
+        }
+
+        private void waitRemoved(ModelControllerClient client, String name) throws Exception {
+            ModelNode address = new ModelNode();
+            address.add("deployment", name);
+            waitStatus(client, "failed", Operations.createOperation("read-resource", address));
+            getLog().debug("Deployment " + name + " removed");
+        }
+
+        private void waitDeploymentUp(ModelControllerClient client, String name) throws Exception {
```

Comment:
I think that we should share the logic with the localDeploymentController, introduce an AbstractDeploymentController?